File: flux-basic-java/src/main/java/com/example/app/user/api/UserCommand.java
package com.example.app.user.api;

import com.example.app.authentication.Sender;
import com.example.app.user.api.model.UserProfile;
import io.fluxzero.sdk.Fluxzero;
import io.fluxzero.sdk.modeling.AssertLegal;
import io.fluxzero.sdk.tracking.TrackSelf;
import io.fluxzero.sdk.tracking.handling.HandleCommand;
import io.fluxzero.sdk.tracking.handling.authentication.UnauthorizedException;
import jakarta.validation.constraints.NotNull;

@TrackSelf
public interface UserCommand {
    @NotNull
    UserId userId();

    @AssertLegal
    default void assertAuthorized(UserProfile user, Sender sender) {
        if (!sender.isAdmin() && !user.userId().equals(sender.userId())) {
            throw new UnauthorizedException("Not authorized to perform this operation");
        }
    }

    @HandleCommand
    default void handle() {
        Fluxzero.loadAggregate(userId()).assertAndApply(this);
    }
}

File: flux-basic-kotlin/src/main/kotlin/com/example/app/user/api/UserCommand.kt
package com.example.app.user.api

import com.example.app.authentication.Sender
import com.example.app.user.api.model.UserProfile
import io.fluxzero.sdk.Fluxzero
import io.fluxzero.sdk.modeling.AssertLegal
import io.fluxzero.sdk.tracking.TrackSelf
import io.fluxzero.sdk.tracking.handling.HandleCommand

@TrackSelf
interface UserCommand {
    val userId: UserId

    @AssertLegal
    fun assertAuthorized(user: UserProfile?, sender: Sender) {
        if (!sender.isAdmin() && user?.userId != sender.userId) {
            throw UserErrors.unauthorized;
        }
    }

    @HandleCommand
    fun handle() {
        Fluxzero.loadAggregate(userId).assertAndApply(this)
    }
}
